### PR TITLE
Shader Refactoring

### PR DESCRIPTION
**Pull request by WestLangley**
Follow-on to #24752.

This post is intended as a basis for further discussion.

EDIT: It looks like we have agreement here. Yay!

### CODE REVIEW

File: src/renderers/shaders/ShaderChunk/envmap_common_pars_fragment.glsl.js [hunk 1]
```diff
@@ -1,7 +1,6 @@
 export default /* glsl */`
 #ifdef USE_ENVMAP
 
-	uniform float backgroundBlurriness;
```
Comment (WestLangley) [inline src/renderers/shaders/ShaderChunk/envmap_common_pars_fragment.glsl.js:4]: IMO, this does not belong here....

File: src/renderers/shaders/ShaderChunk/envmap_fragment.glsl.js [hunk 2]
```diff
@@ -38,10 +38,6 @@ export default /* glsl */`
 
 		vec4 envColor = textureCube( envMap, vec3( flipEnvMap * reflectVec.x, reflectVec.yz ) );
 
-	#elif defined( ENVMAP_TYPE_CUBE_UV )
```
Comment (WestLangley) [inline src/renderers/shaders/ShaderChunk/envmap_fragment.glsl.js:41]: This shader chunk is used only basic, lambert, phong, toon, and gouraud, none of which support ENVMAP_TYPE_CUBE_UV

File: src/renderers/shaders/ShaderLib.js [hunk 3]
```diff
@@ -227,12 +227,11 @@ const ShaderLib = {
...
-			UniformsLib.envmap,
-			{
-				opacity: { value: 1.0 }
-			}
-		] ),
+		uniforms: {
+			tCube: { value: null },
+			tFlip: { value: - 1 },
+			opacity: { value: 1.0 }
+		},
```
Comment (WestLangley) [inline src/renderers/shaders/ShaderLib.js:234]: Restoring the cube shader to its original form.

File: src/renderers/webgl/WebGLBackground.js [hunk 4]
```diff
@@ -68,7 +68,7 @@ function WebGLBackground( renderer, cubemaps, cubeuvmaps, state, objects, alpha,
 					new BoxGeometry( 1, 1, 1 ),
 					new ShaderMaterial( {
 						name: 'BackgroundCubeMaterial',
-						uniforms: cloneUniforms( ShaderLib.cube.uniforms ),
+						uniforms: cloneUniforms( ShaderLib.cube.uniforms ), // todo: use background shader, instead
```
Comment (WestLangley) [inline src/renderers/webgl/WebGLBackground.js:71]: @Mugen87 suggested that perhaps the background shader can be modified to support blurry backgrounds.

**Comment by WestLangley** [comment]
@Mugen87 Some users may be assigning a PMREM as a background. It is fair to say that support for that is no longer needed now that blurry backgrounds are supported?

**Comment by Mugen87** [comment]
Yes, that makes sense. Users can now assign the "raw" or "unprocessed" environment map to `Scene.background` and let the renderer handle the background blurriness.

File: src/renderers/shaders/ShaderChunk.js [hunk 5]
```diff
@@ -106,6 +106,7 @@ import uv2_vertex from './ShaderChunk/uv2_vertex.glsl.js';
 import worldpos_vertex from './ShaderChunk/worldpos_vertex.glsl.js';
 
 import * as background from './ShaderLib/background.glsl.js';
+import * as backgroundCube from './ShaderLib/backgroundCube.glsl.js';
```
Comment (WestLangley) [inline src/renderers/shaders/ShaderChunk.js:109]: Rather that hacking into the existing background shader, I created a new background cube shader that supports PMREM for blurriness.

File: src/renderers/shaders/ShaderLib/backgroundCube.glsl.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+		vec4 envColor = textureCubeUV( envMap, vWorldDirection, backgroundBlurriness );
+
+	#else
+
+		vec4 envColor = vec4( 0.0 );
+
+	#endif
+
+	gl_FragColor = envColor;
+	gl_FragColor.a *= opacity;
```
Comment (WestLangley) [inline src/renderers/shaders/ShaderLib/backgroundCube.glsl.js:44]: @Mugen87 Is opacity support needed here?

**Comment by WestLangley** [comment]
@Mugen87 I am reasonably-happy with this refactoring. Please let me know what you think.

File: src/renderers/shaders/ShaderLib/backgroundCube.glsl.js (hunk 6 above)
Comment (Mugen87) [reply src/renderers/shaders/ShaderLib/backgroundCube.glsl.js:55]: I would say no. We should consider background definitions as opaque.

File: src/renderers/shaders/ShaderChunk.js (hunk 5 above)
Comment (Mugen87) [reply src/renderers/shaders/ShaderChunk.js:109]: I'm okay with this approach. Maybe it makes it easier to maintain the background related shader code.

File: src/renderers/shaders/ShaderLib.js [hunk 7]
```diff
@@ -225,14 +225,27 @@ const ShaderLib = {
 
 	},
 
+	backgroundCube: {
```
Comment (Mugen87) [inline src/renderers/shaders/ShaderLib.js:228]: Instead of using the term `backgroundCube`, what do you think about `skybox` instead?

Comment (WestLangley) [reply src/renderers/shaders/ShaderLib.js:228]: OK, we can change it. How do you feel about `blurredCube` shader? That is the most descriptive name I can think of...

Comment (Mugen87) [reply src/renderers/shaders/ShaderLib.js:228]: True, but I have the feeling `skybox` is what most 3D users understand of what the shader essentially represents. It's a common term in 3D engines.

However, the name is not something that blocks the PR for me. If we can't quickly agree on a name, we can merge the refactoring first and then discuss alternatives. `backgroundCube` isn't wrong after all.

**Review by Mugen87 — APPROVED** [review]

File: src/renderers/shaders/ShaderChunk.js (hunk 5 above)
Comment (WestLangley) [reply src/renderers/shaders/ShaderChunk.js:109]: Yes, that is the hope. :-)

File: src/renderers/shaders/ShaderLib/backgroundCube.glsl.js (hunk 6 above)
Comment (WestLangley) [reply src/renderers/shaders/ShaderLib/backgroundCube.glsl.js]: OK, removed!

File: src/renderers/shaders/ShaderLib.js (hunk 7 above)
Comment (WestLangley) [reply src/renderers/shaders/ShaderLib.js:228]: We already have a skybox shader called 'cube' shader.

- 'cube': a skybox shader, used by post-processing, and other apps

- 'background': used by the renderer for `scene.background` (flat only)

- 'backgroundCube': used by the renderer for blurring `scene.background` (cube only)

The nomenclature is not a deal breaker for me, but let's revisit in a later PR. 😇

**Comment by WestLangley** [comment]
>Some users may be assigning a PMREM as a background. Is it fair to say that support for that is no longer needed now that blurry backgrounds are supported?
>>Yes, that makes sense.

@Mugen87 FYI, with this PR, assigning a PMREM directly to `scene.background` is _still_ supported. Furthermore, simultaneously setting `scene.backgroundBlurriness` will blur it.

I am open leaving that as-is, or removing it in a follow-on PR.

**Comment by Mugen87** [comment]
Let's merge this so we have it in `r146`.

**Comment by WestLangley** [comment]
>Some users may be assigning a PMREM as a background. Is it fair to say that support for that is no longer needed now that blurry backgrounds are supported?

I have changed my mind on this. Some users may have _only_ a PMREM. For example,
```js
scene.environment = pmremGenerator.fromScene( new RoomEnvironment() ).texture;

scene.background = scene.environment; // assign PMREM to background

scene.backgroundBlurriness = 0.2;
```

So, as long as PMREM is used for blurring `scene.background`, I think we should continue to support assigning a PMREM as a scene background texture.

**Comment by Mugen87** [comment]
I've added a note in the migration guide for `r146` about this change. There are some users directly using `THREE.ShaderLib['cube']` and this PR broke their code, see:

https://stackoverflow.com/questions/75149020/property-envmap-removed-from-shader-cube-in-three-js-r146

**Pull request merged**